### PR TITLE
[ENH] explicit default parameter for case_when

### LINKED ISSUE #1159: improvement suggestions for `case_when`
~~- [ ] standalone case_when function, that can be used within assign; this can be handy for multiple case_whens to different column names~~
- [x] More explicit default parameter, currently it is inferred from the variable arguments
- [ ] performance improvement possibly, instead of iterating over all the options, just create matching rows only when needed?

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

~~- standalone function (`if_else`) that can be imported for use in pandas.assign~~
- explicit default parameter added

**This PR resolves #1159 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.md`.

3. [ ] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1165--pyjanitor.netlify.app

File: CHANGELOG.md
Comment (Zeroto521) [inline CHANGELOG.md]: I'm not sure whether you want to use 2-spaces to replace 4-space
If you want to use a 2-spaces style it's better to use cherry-pick this to another PR.

File: janitor/functions/case_when.py
Comment (Zeroto521) [inline janitor/functions/case_when.py]:
```suggestion
    df: pd.DataFrame, *args, *, default: Optional[Any] = 0, column_name: str
```

I suggest using positional argument syntax.
It may avoid some hidden problems since `case_when` could receive `*args`.

File: janitor/functions/case_when.py
Comment (Zeroto521) [inline janitor/functions/case_when.py]: Whether renamed to `_if_else`, `if_else` isn't an accessor.

Comment (Zeroto521) [inline janitor/functions/case_when.py]:
```python
default: Optional[Any]
```
The optional argument needs a default value.
`if_else(df, args)` without `default` will raise an error.

File: janitor/functions/case_when.py
Comment (Zeroto521) [inline janitor/functions/case_when.py]: I prefer the old one. The new one has 4 conditions which are hard to read.

File: janitor/functions/case_when.py
Comment (Zeroto521) [inline janitor/functions/case_when.py]: directly use `is_scalar`?

File: janitor/functions/case_when.py
Comment (Zeroto521) [inline janitor/functions/case_when.py]: use `enumerate` replace `itertools.count`?
So we can remove `import itertools` line.

File: janitor/functions/case_when.py
Comment (Zeroto521) [inline janitor/functions/case_when.py]:
```suggestion
            f"instead got {type(default)!r}"
```

This could output with `''`

**Review by Zeroto521 — CHANGES_REQUESTED** [review]
Small patches need to be fixed.

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: Not sure what you mean here @Zeroto521

Comment (Zeroto521) [reply janitor/functions/case_when.py]: This isn't a big deal. For this long path method, we would like 
to import is_scalar first at the script top. Then use use it directly.
You can ignore this. I'm okay to keep that.

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: this is so that it can be imported and used as a standalone within assign:

```py
from janitor.case_when import if_else
df.assign(a = if else(...), b = if else(...). ...)
```

Of course, if it is not a good idea, we can drop it

File: janitor/functions/case_when.py
Comment (ericmjl) [inline janitor/functions/case_when.py]: @samukweku this is a question more than a critique -- what would make more semantic sense for the default value of the `default` argument? Should it be None/`np.nan` instead of 0? What are your thoughts here? I think you might have given more thought to this problem that I have, so I'm happy to defer to you here.

**Review by ericmjl — COMMENTED** [review]
Just one question from me, @samukweku. Once we resolve it, I'm happy to approve! Other than that, we'll just need to resolve conflicts and then we're good.

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: I am open to criticism :grin: ... I just chose `numpy.select` default value ... thinking about it I'd say None is a better fit. open to suggestions

**Comment by Zeroto521** [comment]
Two threads (https://github.com/pyjanitor-devs/pyjanitor/runs/8295244245, https://github.com/pyjanitor-devs/pyjanitor/runs/8295231459) CI failed due to new PR #1158.

`case_when` testcases raised some errors like 

```python
=========================== short test summary info ============================
ERROR tests/functions/test_case_when.py - hypothesis.errors.InvalidArgument: ...
ERROR tests/functions/test_case_when.py - hypothesis.errors.InvalidArgument: ...
ERROR tests/functions/test_case_when.py - hypothesis.errors.InvalidArgument: ...
ERROR tests/functions/test_case_when.py - hypothesis.errors.InvalidArgument: ...
```

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: I'll edit and let the user pass a default value instead

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: since args is already variable arg, this is fine, and no issues should arise

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: dropping the `if_else` idea altogether; user can do multiple case_when if necessary, similar to what obtains in SQL. In the future, users can change this, if there is enough interest

File: janitor/functions/case_when.py
Comment (thatlittleboy) [inline janitor/functions/case_when.py]: My only concern is that this is a breaking change which probably warrants a deprecation cycle: users who depended on providing `2n+1` args (with the last arg treated as the "default"), will now be met with a TypeError due to a missing `default` kwarg.

I'm thinking if a better solution here is to (1) make `default` default to some value like `"__pyjanitor_default__"` (it can't be None since I suppose None is a valid default value), (2) introduce a DeprecationWarning to when we detect `default` is the special value from step1. Point the user towards using the explicit `default` kwarg. while manually popping the last arg into `default`. (3) Perform the remaining checks as if the user had passed in `default`, i.e., if `len(args)` is odd, then raise error.

Thoughts?

**Review by thatlittleboy — CHANGES_REQUESTED** [review]
Just 1 concern on my end w.r.t. breaking changes. Hope we can address it before we merge. Thanks

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: Been a while @thatlittleboy hope you are well and good. 

I'm thinking of closing this PR and leave things as they are; the core is not really changed, maybe a future PR can improve performance. But yea, the breaking change is not worth it

Comment (thatlittleboy) [reply janitor/functions/case_when.py]: Hey @samukweku , thanks, it's been a while. Hoping to jump back in some time soon to assist more, probably during the holiday season when work's a little lighter.

Back on topic: I do like the idea of specifying the `default` kwarg though, I think it's a step in the right direction. I've always found the (2n+1) "API" for `case_when` a little confusing to read when written out as code. The breaking change just needs a little careful handling.

I'll suggest keeping the PR open instead of closing for now, someone else can pick up from where you left off (I'll volunteer, if I can find the time :x ).

Comment (samukweku) [reply janitor/functions/case_when.py]: @thatlittleboy made some edits to the code to cater for deprecation. Would love your feedback on this. @ericmjl @Zeroto521 pls feel free to chime in when you have some bandwidth

File: tests/functions/test_case_when.py
Comment (thatlittleboy) [inline tests/functions/test_case_when.py]: I think it's more important to ensure in this test that `DeprecationWarning` is raised when the user is writing in the style from before. i.e., I suggest we remove the `default=None`.

File: janitor/functions/case_when.py
Comment (thatlittleboy) [inline janitor/functions/case_when.py]: Just curious why we are raising and error rather than converting the list/tuple/etc. to a Series (as we've done before)?

**Review by thatlittleboy — APPROVED** [review]
Thanks Sam, just a couple questions left from me, but nothing major. Pre-approving.

smiling-cat-face-lgtm

File: tests/functions/test_case_when.py
Comment (samukweku) [reply tests/functions/test_case_when.py]: @thatlittleboy Do you mind scribbling what U mean so I can better understand?

File: janitor/functions/case_when.py
Comment (samukweku) [reply janitor/functions/case_when.py]: Good catch. Lemme fix

File: tests/functions/test_case_when.py
Comment (thatlittleboy) [reply tests/functions/test_case_when.py]:
```python
    with pytest.warns(DeprecationWarning):
        dataframe.case_when(
            dataframe.a < 10,
            "less_than_10",
            dataframe.a == 5,
            column_name="a",
        )
```
I just mean this ^ :)

(When I said "style from before", I meant "before this PR / before the breaking change is introduced".)

**Comment by ericmjl** [comment]
@thatlittleboy would you like to do the honors here? a -0.01% decrease in coverage is marginal, we can handle it :smile:

**Comment by thatlittleboy** [comment]
yep okies, merging

**Pull request merged**
